### PR TITLE
Some additions

### PR DESCRIPTION
**Pull request by phadej**
I got nerd-snippet by whether we can have better type for `step` question.

### CODE REVIEW

**Comment by phadej** [comment]
Last commit (about Sing (Idx n)) is a puzzle for @RyanGlScott 

EDIT: I see that is discussed in #2; EDIT2: I'm optimistic that SubstScoped can be promoted.

**Comment by RyanGlScott** [comment]
As you've already noted in the edit to your comment in https://github.com/sweirich/challenge/pull/4#issuecomment-648475236, I once tried to solve the puzzle of automating everything in `SubstScoped` [and failed](https://github.com/sweirich/challenge/issues/2#issuecomment-647707940). Perhaps you will have better luck?

**Comment by phadej** [comment]
Oh no, *automating* is difficult when we don't even know what we want to be generated.

Yet with

```diff
diff --git a/debruijn/src/SubstScoped.hs b/debruijn/src/SubstScoped.hs
index afca123..335b733 100644
--- a/debruijn/src/SubstScoped.hs
+++ b/debruijn/src/SubstScoped.hs
@@ -20,9 +20,9 @@ data Idx :: Nat -> Type where
 -- We can lift Idx, but singletons cannot do that for us :/
 data SIdx (i :: Idx n) where
     SFZ :: SIdx 'FZ
-    SFS :: SIdx i -> SIdx ('FS i)
+    SFS :: SIdx (i :: Idx n) -> SIdx ('FS i :: Idx ('S n))
 
-type instance Sing = SIdx
+type instance Sing = SIdx :: Idx n -> Type
 
 instance SingI 'FZ where sing = SFZ
 instance SingI i => SingI ('FS i) where sing = SFS sing
@@ -66,14 +66,26 @@ instance C.Category (Sub a) where
     id = nilSub
     (.) = flip (:<>)
 
+data SSub (sub :: Sub a n m) where
+    SInc   :: forall m (sm :: Sing m). Sing m -> SSub ('Inc sm)
+    (:%<)  :: Sing a -> SSub sub -> SSub (a ':< sub)
+    (:%<>) :: SSub sub1 ->SSub sub2 ->SSub (sub1 ':<> sub2)
+
 -- | Identity substitution, leaves all variables alone
 nilSub :: Sub a n n 
 nilSub = Inc SZ
 
+type NilSub = 'Inc 'SZ :: Sub a n n
+
+sNilSub :: SSub NilSub
+sNilSub = SInc SZ
+
 -- | Increment, shifts all variables by one
 incSub :: Sub a n (S n)
 incSub = Inc (SS SZ)
 
+type IncSub = 'Inc ('SS 'SZ) :: Sub a n (S n)
+
 -- | Singleton, replace 0 with t, leave everything else alone
 singleSub :: a n -> Sub a (S n) n
 singleSub t = t :< nilSub
@@ -89,6 +101,14 @@ add :: Sing m -> Idx n -> Idx (Plus m n)
 add SZ x = x
 add (SS m) x = FS (add m x)
 
+type family Add (m :: Nat) (i :: Idx n) :: Idx (Plus m n) where
+    Add 'Z     x = x
+    Add ('S m) x = 'FS (Add m x)
+
+sAdd :: Sing m -> Sing i -> Sing (Add m i)
+sAdd SZ     x = x
+sAdd (SS m) x = SFS (sAdd m x)
+
 -- | Value of the index x in the substitution s
 applySub :: SubstDB a => Sub a n m -> Idx n -> a m
 applySub (Inc m)        x  = var (add m x)
```

only `SubstDB` and `applySub` (I consider `lift` easy after that) need promotion. There one need some defunctionalization term for `SubstDB` methods, but I think it is doable.

Here I chose (= the first thing I tried) is to avoid double promotion of singletons. It seems to work so far.

**Comment by RyanGlScott** [comment]
Certainly, I have no doubt that you can manually promote and single everything in `SubstScoped` if you so chose. I have to imagine that will be quite painful, however. :)

**Comment by phadej** [comment]
@RyanGlScott I got quite far, cannot give up at this point.

We have weird hobbies.

**Review by sweirich — CHANGES_REQUESTED** [review]
Thanks for the addition! This is fun.

Can you put this to a separate file, similar to Cps.hs and Typecheck.hs? I'd like to keep the main 'SimpleTyped' module short. (And maintain parity between it and the other versions (Poly/Simple/PolyTyped/...)

**Review by sweirich — COMMENTED** [review]
I've created a separate branch for experiments with singletons for Idx (i.e. adding type scopes).

**Review by sweirich — APPROVED** [review]

File: debruijn/src/SimpleTyped.hs [hunk 1]
```diff
@@ -101,4 +101,115 @@ reduce (AppE (LamE t e1) e2)   = subst (singleSub (reduce e2)) (reduce e1)
...
+-- again, (almost) the same instance definition as before.
+instance SubstDB Norm where
+   var = Norm . NeutN . VarN
+
+   subst s (Norm (IntN x))         = Norm (IntN x)
+   subst s (Norm (LamN ty e))      = Norm (LamN ty (subst (lift s) e))
+   subst s (Norm (NeutN (VarN x))) = applySub s x
+
+   -- substitution in application causes evaluation,
+   -- i.e. recursive substitutions.
```
Comment (sweirich) [inline debruijn/src/SimpleTyped.hs]: Awesome! I didn't realize that hereditary substitutions work with this interface too.

File: debruijn/src/SimpleTyped.hs [hunk 2]
```diff
@@ -101,4 +101,115 @@ reduce (AppE (LamE t e1) e2)   = subst (singleSub (reduce e2)) (reduce e1)
...
+
+-- Yet another variant of 'Exp'.
+--
+-- How handy that Neutral and Normal start with letter N.
+--
+newtype Norm g t = Norm (GenNorm Norm g t)
+type Neut = GenNeut Norm
+
+-- We define more generic version of 'Norm' and 'Neut',
+-- so we can define
```
Comment (sweirich) [inline debruijn/src/SimpleTyped.hs]: did you mean to put "WHNF" at the end of this sentence?

File: debruijn/src/SimpleTyped.hs [hunk 3]
```diff
@@ -101,4 +101,115 @@ reduce (AppE (LamE t e1) e2)   = subst (singleSub (reduce e2)) (reduce e1)
...
+      -- The following is easiest without any additional machinery,
+      -- substitute terms as is. This most likely duplicates work.
+      --
+      --   reduceN (subst (singleSub e2) e1')
+      --
+      -- or we can do
+      --
+      --  reduceN (subst (singleSub (norm2exp (reduceN e2))) (norm2exp (reduceN e1')))
+      --  ^ I think the reduceN after subst is missing from 'reduce' on 'Exp'.
+      --
```
Comment (sweirich) [inline debruijn/src/SimpleTyped.hs]: Thanks!  That example was indeed confused.  I've just updated the example to be parallel reduction (which doesn't necessarily produce a normal form in one step, but does demonstrate reduction under binders.)

**Comment by phadej** [comment]
I split `Norm` things into own module. Also removed `Idx` changes from this branch.

**Review by sweirich — CHANGES_REQUESTED** [review]
Thanks.  Can you also move the by-hand singletons in SubstScoped to the other branch?  I don't want to give the impression that they are necessary for this example.

**Comment by phadej** [comment]
> Thanks. Can you also move the by-hand singletons in SubstScoped to the other branch? I don't want to give the impression that they are necessary for this example.

That's done.

**Review by sweirich — APPROVED** [review]
Thanks!

**Pull request merged**